### PR TITLE
Try to fix a template issue on `cucumber-jvm-migration`

### PR DESCRIPTION
**Pull request by kunli2**
This PR attempts to fix the issue https://github.com/openrewrite/rewrite/issues/3102
I think the issue was introduced by [this commit](https://github.com/openrewrite/rewrite/commit/626d8c645bb7b6b9d81e69a7985759701eb9b133#r108541429), 
This PR just patched against that specified problem, it can fix the template problem, but it's a patch fix. I don't know whether it's a correct change.  I think the issue needs @knutwannheden 's help to also take a look.

### CODE REVIEW

**Review by sambsnyd — APPROVED** [review]

File: rewrite-java/src/main/java/org/openrewrite/java/internal/template/BlockStatementTemplateGenerator.java [hunk 1]
```diff
@@ -264,7 +264,7 @@ private void template(Cursor cursor, J prior, StringBuilder before, StringBuilde
                 before.insert(0, "{\n");
             }
 
-            if (prior instanceof Statement) {
+            if (prior instanceof Statement && !(prior instanceof J.MethodDeclaration)) {
                 after.append(';');
             }
```
Comment (sambsnyd) [inline rewrite-java/src/main/java/org/openrewrite/java/internal/template/BlockStatementTemplateGenerator.java:269]: `J.MethodDeclaration` and `J.ClassDeclaration` should not have a `;` appended, although it is generally not a syntax error to have one. (Does result in an extra empty statement being generated). 

There may be other exceptions to the rule, but those are the two that occur to me thinking about it.

Comment (shanman190) [reply rewrite-java/src/main/java/org/openrewrite/java/internal/template/BlockStatementTemplateGenerator.java:269]: A number of control flow statements also come to mind as well (if, for, while, and switch to name a few)

Comment (knutwannheden) [reply rewrite-java/src/main/java/org/openrewrite/java/internal/template/BlockStatementTemplateGenerator.java:269]: I was the one to add a semicolon after statements here. The problem as I remember it has to do with the fact that we have some expressions which can also be used as statements (e.g. method invocations) and unlike the Javac AST we don't wrap them in some kind of `ExpressionStatement` element.

I will take a look at this, add some more tests and possibly add a cleaner fix.

**Comment by knutwannheden** [comment]
There is many aspects to this problem and all of them put together produce this behavior:
- Cucumber uses a JavaTemplate which basically looks like this: `void m() { #{any()} }`. Since `#{any()}` basically represents an expression the template really should include a terminating semicolon, as it is used as a statement.
- The `javac` parser is really quite permissive when constructing the AST and is often able to ignore missing semicolons and other minor syntax errors
- So even though the `javac` compiler successfully produces a "correct" AST for the template above (which for example could end up being translated to `/*__TEMPLATE__*/void m() { __P__./*__p0__*/intp() }/*__TEMPLATE_STOP__*/;` our Java parser then associates the `/*__TEMPLATE_STOP__*/` comment with the templated `p0` statement, because it scans forward until it finds the terminating semicolon.
- Whereas when we don't insert a semicolon after the method declaration, then our Java parser associates the comment with the method declaration (using `J.RightPadded`).

I agree that we should be removing the semicolon in this context (I will make a proposal). At the same time I think it would be good for Cucumber to also add a terminating semicolon to their templates, so that the Java source file being parsed is syntactically correct.

**Comment by knutwannheden** [comment]
I slightly changed the fix, added a test case, and also corrected two similar templates we had in our code.

**Comment by kunli2** [comment]
nice, Thanks @knutwannheden for looking into this!

**Pull request merged**